plugins/catalog-backend/src/catalog/DatabaseCatalog.ts

```diff
@@ -15,29 +15,79 @@
  */
 
 import { NotFoundError } from '@backstage/backend-common';
-import path from 'path';
 import Knex from 'knex';
+import path from 'path';
+import fs from 'fs-extra';
 import { v4 as uuidv4 } from 'uuid';
-import { AddLocationRequest, Component, Catalog, Location } from './types';
+import { Logger } from 'winston';
+import {
+  AddLocationRequest,
+  Catalog,
+  Component,
+  Location,
+  componentShape,
+} from './types';
+import YAML from 'yaml';
+
+async function readLocation(location: Location): Promise<Component[]> {
+  switch (location.type) {
+    case 'file': {
+      let parsed;
+      try {
+        const raw = await fs.readFile(location.target, 'utf8');
+        parsed = YAML.parse(raw);
+      } catch (e) {
+        throw new Error(`Unable to read "${location.target}", ${e}`);
+      }
+
+      try {
+        return [await componentShape.validate(parsed, { strict: true })];
+      } catch (e) {
+        throw new Error(
+          `Malformed file contents at "${location.target}", ${e}`,
+        );
+      }
+    }
+
+    default:
+      throw new Error(`Unknown type "${location.type}"`);
+  }
+}
 
 export class DatabaseCatalog implements Catalog {
-  static async create(database: Knex): Promise<DatabaseCatalog> {
+  static async create(
+    database: Knex,
+    logger: Logger,
+  ): Promise<DatabaseCatalog> {
     await database.migrate.latest({
       directory: path.resolve(__dirname, '..', 'migrations'),
       loadExtensions: ['.js'],
     });
 
-    return new DatabaseCatalog(database);
+    const databaseCatalog = new DatabaseCatalog(database, logger);
+
+    const startRefresh = async () => {
```

Comment:
I prefer these things outside the constructor, in a start/stop, or open/close pair of methods. Makes it easier to test and lets you check asynchronous pre-work.